statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,370 @@
+		idxMem := tblMemUsage.IndicesMemUsage[idxID]
+		s.lru.put(tblID, idxID, idx, idxMem, true, needMove)
+	}
+}
+
+// Del implements statsCacheInner
+func (s *statsInnerCache) Del(tblID int64) {
+	s.Lock()
+	defer s.Unlock()
+	element, exist := s.elements[tblID]
+	if !exist {
+		return
+	}
+	// remove indices
+	for idxID := range element.tbl.Indices {
+		s.lru.del(tblID, idxID)
+	}
+	delete(s.elements, tblID)
+}
+
+// Cost implements statsCacheInner
+func (s *statsInnerCache) Cost() int64 {
+	s.RLock()
+	defer s.RUnlock()
+	return s.lru.trackingCost
+}
+
+func (s *statsInnerCache) totalCost() int64 {
+	s.RLock()
+	defer s.RUnlock()
+	totalCost := int64(0)
+	for tblID, ele := range s.elements {
+		s.freshTableCost(tblID, ele)
+		totalCost += ele.tblMemUsage.TotalMemUsage
+	}
+	return totalCost
+}
+
+// Keys implements statsCacheInner
+func (s *statsInnerCache) Keys() []int64 {
+	s.RLock()
+	defer s.RUnlock()
+	r := make([]int64, 0, len(s.elements))
+	for tblID := range s.elements {
+		r = append(r, tblID)
+	}
+	return r
+}
+
+// Values implements statsCacheInner
+func (s *statsInnerCache) Values() []*statistics.Table {
+	s.RLock()
+	defer s.RUnlock()
+	r := make([]*statistics.Table, 0, len(s.elements))
+	for _, v := range s.elements {
+		r = append(r, v.tbl)
+	}
+	return r
+}
+
+// Map implements statsCacheInner
+func (s *statsInnerCache) Map() map[int64]*statistics.Table {
+	s.RLock()
+	defer s.RUnlock()
+	r := make(map[int64]*statistics.Table, len(s.elements))
+	for k, v := range s.elements {
+		r[k] = v.tbl
+	}
+	return r
+}
+
+// Len implements statsCacheInner
+func (s *statsInnerCache) Len() int {
+	s.RLock()
+	defer s.RUnlock()
+	return len(s.elements)
+}
+
+// FreshMemUsage implements statsCacheInner
+func (s *statsInnerCache) FreshMemUsage() {
+	s.Lock()
+	defer s.Unlock()
+	for tblID, element := range s.elements {
+		s.freshTableCost(tblID, element)
+	}
+}
+
+// FreshTableCost implements statsCacheInner
+func (s *statsInnerCache) FreshTableCost(tblID int64) {
+	s.Lock()
+	defer s.Unlock()
+	element, exist := s.elements[tblID]
+	if !exist {
+		return
+	}
+	s.freshTableCost(tblID, element)
+}
+
+// Copy implements statsCacheInner
+func (s *statsInnerCache) Copy() statsCacheInner {
+	s.RLock()
+	defer s.RUnlock()
+	newCache := newStatsLruCache(s.lru.capacity)
+	newCache.lru = s.lru.copy()
+	for tblID, element := range s.elements {
+		newCache.elements[tblID] = element
+	}
+	return newCache
+}
+
+func (s *statsInnerCache) onEvict(tblID int64) {
+	element, exist := s.elements[tblID]
+	if !exist {
+		return
+	}
+	element.tblMemUsage = element.tbl.MemoryUsage()
+}
+
+func (s *statsInnerCache) freshTableCost(tblID int64, element *lruMapElement) {
+	newTblMem := element.tbl.MemoryUsage()
+	for idxID, idx := range element.tbl.Indices {
+		s.lru.put(tblID, idxID, idx, newTblMem.IndicesMemUsage[idxID], true, false)
+	}
+	// tbl mem usage might be changed due to evict
+	element.tblMemUsage = element.tbl.MemoryUsage()
```

Comment:
Is it duplicated with element.tbl.MemoryUsage() which is called in onEvict?

Reply:
right, it's unnecessary now, updated.